server/src/main/java/org/opensearch/index/mapper/CompositeMappedFieldType.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.mapper;
+
+import org.opensearch.common.annotation.ExperimentalApi;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Base class for composite field types
+ *
+ * @opensearch.experimental
+ */
+@ExperimentalApi
+public abstract class CompositeMappedFieldType extends MappedFieldType {
+    private final List<String> fields;
+    private final CompositeFieldType type;
+
+    public CompositeMappedFieldType(
+        String name,
+        boolean isIndexed,
+        boolean isStored,
+        boolean hasDocValues,
+        TextSearchInfo textSearchInfo,
+        Map<String, String> meta,
+        List<String> fields,
+        CompositeFieldType type
+    ) {
+        super(name, isIndexed, isStored, hasDocValues, textSearchInfo, meta);
+        this.fields = fields;
+        this.type = type;
+    }
+
+    public CompositeMappedFieldType(String name, List<String> fields, CompositeFieldType type) {
+        this(name, false, false, false, TextSearchInfo.NONE, Collections.emptyMap(), fields, type);
+    }
+
+    /**
+     * Supported composite field types
+     *
+     * @opensearch.experimental
+     */
+    @ExperimentalApi
+    public enum CompositeFieldType {
+        STAR_TREE("star_tree");
```

Comment:
nit: rename better, maybe data cube

Reply:
The mapping [changes](https://github.com/opensearch-project/OpenSearch/pull/14261/files) has a hierarchy - compositeMappedFieldType -> DataCubeMappedFieldType -> StarTreeMappedFieldType .. Here I've just picked this particular class to use the associated methods.